pumpkin-inventory/src/lib.rs

```diff
@@ -52,3 +62,185 @@ impl WindowType {
         "WINDOW TITLE"
     }
 }
+// Container needs Sync + Send to be able to be in async Server
+pub trait Container: Sync + Send {
+    fn window_type(&self) -> &'static WindowType;
+
+    fn handle_item_change(
+        &mut self,
+        carried_item: &mut Option<ItemStack>,
+        slot: usize,
+        mouse_click: MouseClick,
+    ) -> Result<(), InventoryError> {
+        let mut all_slots = self.all_slots();
+        if slot > all_slots.len() {
+            Err(InventoryError::InvalidSlot)?
+        }
+        handle_item_change(carried_item, all_slots[slot], mouse_click);
+        Ok(())
+    }
+
+    fn all_slots(&mut self) -> Vec<&mut Option<ItemStack>>;
+
+    fn all_slots_ref(&self) -> Vec<Option<&ItemStack>>;
+
+    fn all_combinable_slots(&self) -> Vec<Option<&ItemStack>> {
+        self.all_slots_ref()
+    }
+
+    fn all_combinable_slots_mut(&mut self) -> Vec<&mut Option<ItemStack>> {
+        self.all_slots()
+    }
+
+    fn print_all_contents(&self) {
+        self.all_slots_ref()
+            .into_iter()
+            .enumerate()
+            .filter_map(|(slot, item)| item.map(|item| (slot, item)))
+            .for_each(|(slot, item)| {
+                dbg!(slot, item);
+            });
+    }
+
+    fn internal_pumpkin_id(&self) -> u64 {
+        0
+    }
+}
+
+pub fn handle_item_take(
+    carried_item: &mut Option<ItemStack>,
+    item_slot: &mut Option<ItemStack>,
+    mouse_click: MouseClick,
+) {
+    let Some(item) = item_slot else {
+        return;
+    };
+    let mut new_item = *item;
+
+    match mouse_click {
+        MouseClick::Left => {
+            *item_slot = None;
+        }
+        MouseClick::Right => {
+            let half = item.item_count / 2;
+            item.item_count -= half;
+            new_item.item_count = half;
+        }
+    }
+    *carried_item = Some(new_item);
+}
```

Comment:
This looks like an not rusty approche. I you can just get not mut input values and then return the the value. Something like
```rust
pub fn handle_item_take(
    carried_item: Option<ItemStack>,
    item_slot: Option<ItemStack>,
    mouse_click: MouseClick,
) -> (Option<ItemStack>, Option<ItemStack>) {
    let Some(item) = item_slot else {
        return (None, None);
    };
   let item_slot = item_slot;
    match mouse_click {
        MouseClick::Left => {
            *item_slot = None;
        }
        MouseClick::Right => {
            let half = item.item_count / 2;
            item.item_count -= half;
            new_item.item_count = half;
        }
    }
    (new_item, item_slot);
}
```

Reply:
I feel like it is a good approach, both `&mut Option<ItemStack>` are inputs and outputs, and it is necessary for them to switch anyway. I don't think this is not a rusty approach, since the situation warrants this kind of api, I mean in rust you already need to mark something mut, so I also don't think the usage of this function would be annoying.

I do think however that there shoudld be a clear code doc to explain this behaviour.